modules/oe_editorial_entity_version/tests/src/Functional/NodeRevisionRevertTest.php

```diff
@@ -0,0 +1,149 @@
+    'oe_editorial_entity_version',
+  ];
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setUp() {
+    parent::setUp();
+
+    $entity_type_manager = $this->container->get('entity_type.manager');
+    $this->nodeStorage = $entity_type_manager->getStorage('node');
+    $this->node = $this->nodeStorage->create(
+      [
+        'type' => 'oe_workflow_demo',
+        'title' => 'My node',
+        'moderation_state' => 'draft',
+      ]
+    );
+    $this->node->save();
+
+    /** @var \Drupal\user\RoleInterface $role */
+    $role = $entity_type_manager->getStorage('user_role')->load('oe_validator');
+    $permissions = $role->getPermissions();
+    $permissions[] = 'administer nodes';
+    $this->user = $this->drupalCreateUser($permissions);
+    $this->drupalLogin($this->user);
+  }
+
+  /**
+   * Tests the node revision revert confirm form.
+   */
+  public function testNodeRevisionRevertConfirmForm(): void {
+    $initial_revision_id = $this->node->getRevisionId();
+    $initial_version = [
+      $this->node->version->major,
+      $this->node->version->minor,
+      $this->node->version->patch,
+    ];
+    $initial_revision_date = $this->node->getRevisionCreationTime();
+
+    // Now create a new major version with a different title.
+    $this->node->setTitle('My node ready to publish');
+    $this->node->moderation_state->value = 'request_validation';
+    $this->node->save();
+    $this->node->moderation_state->value = 'validated';
+    $this->node->save();
+    // Now Publish the node.
+    $this->node->moderation_state->value = 'published';
+    $this->node->save();
+
+    // Revert the node to the initial revision.
+    $revert_url = Url::fromRoute('node.revision_revert_confirm', [
+      'node' => $this->node->id(),
+      'node_revision' => $initial_revision_id,
+    ]);
+
+    // Assert that the user can't access the url because lacks permission.
+    $this->assertFalse($revert_url->access($this->user));
+
+    // Grant the 'restore version' permission to the user.
+    $entity_type_manager = $this->container->get('entity_type.manager');
+    $role = $entity_type_manager->getStorage('user_role')->load($this->user->getRoles(TRUE)[0]);
+    $role->grantPermission('restore version');
+    $role->save();
+
+    // Assert that the user can access the url.
+    $this->assertTrue($revert_url->access($this->user));
+
+    $this->drupalGet($revert_url);
+
+    $date_formatter = $this->container->get('date.formatter');
+    // Assert we are on the correct page with the correct confirm message.
+    $this->assertSession()->pageTextContains('Are you sure you want to restore the ' . implode('.', $initial_version) . ' version from ' . $date_formatter->format($initial_revision_date) . '?');
+    // A cancel button is present.
+    $this->assertSession()->linkExists('Cancel');
+    // Assert the restore button exist and restore the version by pressing it.
+    $this->assertSession()->buttonExists('Restore')->press();
```

Comment:
A version can be restored from any moderation state. Can you please test the restore works properly from each state?

Reply:
The actual state of the revision does not change the logic, or the ability to restore, that's why it is not part of the tests, because the same thing happens in all states.